File: mcp-code-intelligence-nodejs/src/http/viewer-server.ts
/**
 * HTTP server for Knowledge Graph viewer — uses Node.js built-in http module.
 * All HTML/CSS/JS served from shared/viewer/ (single source of truth).
 */

import * as http from 'http';
import * as fs from 'fs';
import * as path from 'path';
import { MemoryEngine } from '../memory/memory-engine.js';
import { KnowledgeGraph } from '../memory/knowledge-graph.js';
import { handleApiRoute } from './api-routes.js';
import { handleIngestFileRoute } from './ingest-routes.js';
import { handleUxRoute } from './ux-routes.js';
import { handleModelRoute } from './model-routes.js';
import { ModelManager } from '../orchestration/models/model-manager.js';

export class ViewerServer {
  private server: http.Server | null = null;
  private port: number;

  /** Late-binding — set after MCP initialize completes. */
  memoryEngine: MemoryEngine | null = null;
  knowledgeGraph: KnowledgeGraph | null = null;
  modelManager: ModelManager | null = null;
  workspace: string = '';

  constructor(port: number, workspace: string = '') {
    this.port = port;
    this.workspace = workspace;
  }

  start(): void {
    this.server = http.createServer((req, res) => this.handleRequest(req, res));
    this.server.listen(this.port, () => {
      const addr = this.server!.address() as any;
      const actualPort = addr?.port ?? this.port;
      this.port = actualPort;
      console.error(`[code-intel] HTTP viewer starting on port ${actualPort}`);
    });
  }

  stop(): void {
    this.server?.close();
    this.server = null;
  }

  private handleRequest(req: http.IncomingMessage, res: http.ServerResponse): void {
    const url = new URL(req.url ?? '/', `http://localhost:${this.port}`);
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    if (req.method === 'OPTIONS') { res.writeHead(204); res.end(); return; }

    if (url.pathname === '/') {
      this.serveSharedFile(res, 'index.html', 'text/html');
    } else if (url.pathname === '/dashboard') {
      this.serveSharedFile(res, 'dashboard.html', 'text/html');
    } else if (url.pathname === '/tags') {
      this.serveSharedFile(res, 'tags.html', 'text/html');
    } else if (url.pathname === '/quality') {
      this.serveSharedFile(res, 'quality.html', 'text/html');
    } else if (url.pathname === '/analytics') {
      this.serveSharedFile(res, 'analytics.html', 'text/html');
    } else if (url.pathname.startsWith('/modules/') || url.pathname.startsWith('/config/')) {
      this.serveSubdirStatic(url.pathname, res);
    } else if (url.pathname.endsWith('.js') || url.pathname.endsWith('.css')) {
      this.serveStatic(url.pathname, res);
    } else if (url.pathname === '/api/health') {
      this.serveHealth(res);
    } else if (url.pathname.startsWith('/api/models')) {
      if (!handleModelRoute(req, url, res, this.modelManager)) {
        this.send404(res);
      }
    } else if (url.pathname.startsWith('/api/kb')) {
      const db = this.memoryEngine?.db ?? null;
      if (!handleUxRoute(req, url, res, this.memoryEngine, db)) {
        this.send404(res);
      }
    } else if (url.pathname.startsWith('/api/memory')) {
      if (req.method === 'POST') {
        handleIngestFileRoute(req, url, res, this.memoryEngine, this.workspace);
      } else {
        handleApiRoute(url, res, this.memoryEngine, this.knowledgeGraph);
      }
    } else {
      this.send404(res);
    }
  }

  private serveSharedFile(res: http.ServerResponse, filename: string, contentType: string): void {
    const filePath = this.resolveSharedPath(filename);
    if (!filePath) { this.serveViewerError(res, filename); return; }
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      res.writeHead(200, { 'Content-Type': `${contentType}; charset=utf-8` });
      res.end(content);
    } catch { this.serveViewerError(res, filename); }
  }

  private serveSubdirStatic(pathname: string, res: http.ServerResponse): void {
    if (pathname.includes('..')) { this.send404(res); return; }
    const relPath = pathname.startsWith('/') ? pathname.slice(1) : pathname;
    const filePath = this.resolveSharedPath(relPath);
    if (!filePath) { this.send404(res); return; }
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const ct = pathname.endsWith('.json') ? 'application/json'
        : pathname.endsWith('.css') ? 'text/css' : 'application/javascript';
      res.writeHead(200, { 'Content-Type': ct + '; charset=utf-8' });
      res.end(content);
    } catch { this.send404(res); }
  }

  private serveStatic(pathname: string, res: http.ServerResponse): void {
    const filename = path.basename(pathname);
    if (filename.includes('..')) { this.send404(res); return; }
    const filePath = this.resolveSharedPath(filename);
    if (!filePath) { this.send404(res); return; }
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const ct = filename.endsWith('.css') ? 'text/css' : 'application/javascript';
      res.writeHead(200, {
        'Content-Type': ct + '; charset=utf-8',
        'Cache-Control': 'no-cache, no-store, must-revalidate',
        'Pragma': 'no-cache',
      });
      res.end(content);
    } catch { this.send404(res); }
  }

  /** Resolve path within viewer/. Checks bundled dist/viewer/ first, then workspace shared/viewer/. */
  private resolveSharedPath(relPath: string): string | null {
    // 1. Bundled viewer (inside npm package: dist/viewer/)
    const bundledPath = path.join(__dirname, '..', 'viewer', relPath);
    if (fs.existsSync(bundledPath)) return bundledPath;
    // 2. Fallback: workspace shared/viewer/ (dev mode)
    if (this.workspace) {
      const wsPath = path.join(this.workspace, 'shared', 'viewer', relPath);
      if (fs.existsSync(wsPath)) return wsPath;
    }
    return null;
  }

  /** Error page when shared/viewer/ files are missing. */
  private serveViewerError(res: http.ServerResponse, filename: string): void {
    const html = `<!DOCTYPE html><html><head><title>Viewer Unavailable</title></head>`
      + `<body style="background:#0f172a;color:#e2e8f0;font-family:system-ui;padding:2rem">`
      + `<h1>Viewer Unavailable</h1>`
      + `<p>shared/viewer/${filename} not found. Please ensure workspace is correct.</p>`
      + `<p style="opacity:.6;font-size:.8rem">Workspace: ${this.workspace || '(not set)'}</p>`
      + `</body></html>`;
    res.writeHead(503, { 'Content-Type': 'text/html; charset=utf-8' });
    res.end(html);
  }

  private serveHealth(res: http.ServerResponse): void {
    this.sendJson(res, {
      status: 'ok',
      version: '0.2.0',
      workspace: this.workspace,
      viewerPort: this.port,
      memoryEnabled: this.memoryEngine !== null,
    });
  }

  private send404(res: http.ServerResponse): void {
    res.writeHead(404, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'Not found' }));
  }

  private sendJson(res: http.ServerResponse, data: unknown): void {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(data));
  }
}

File: mcp-code-intelligence-python/src/mcp_code_intel/http/viewer_server.py
"""HTTP server for Knowledge Graph viewer — uses Python stdlib.
All HTML/CSS/JS served from shared/viewer/ (single source of truth).
"""

import os
import sys
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs

from .api_routes import handle_api_route, handle_health
from .ingest_routes import handle_ingest_file_route
from .kb_viewer_routes import handle_kb_viewer_route
from .model_routes import handle_model_route
from .ux_routes import handle_ux_route


class ViewerServer:
    """Serves REST API + Knowledge Graph Web Viewer."""

    def __init__(self, port: int, workspace: str = "") -> None:
        self.port = port
        self.workspace = workspace
        self.memory_engine = None
        self.knowledge_graph = None
        self.model_manager = None
        self._server: HTTPServer | None = None
        self._thread: threading.Thread | None = None

    def start(self) -> None:
        """Start HTTP server in daemon thread."""
        server_ref = self

        class Handler(BaseHTTPRequestHandler):
            def do_GET(self) -> None:
                _route_get(self, server_ref)

            def do_POST(self) -> None:
                _route_post(self, server_ref)

            def do_OPTIONS(self) -> None:
                _handle_cors_preflight(self)

            def log_message(self, format: str, *args: object) -> None:
                pass

        self._server = HTTPServer(("0.0.0.0", self.port), Handler)
        self._server.viewer_workspace = self.workspace
        self._thread = threading.Thread(
            target=self._server.serve_forever, daemon=True
        )
        self._thread.start()
        _log(f"Viewer server started on http://localhost:{self.port}")

    def stop(self) -> None:
        """Stop the server."""
        if self._server:
            self._server.shutdown()
            self._server = None
            self._thread = None
            _log("Viewer server stopped")


_PAGE_MAP = {
    "/": "index.html",
    "/dashboard": "dashboard.html",
    "/tags": "tags.html",
    "/quality": "quality.html",
    "/analytics": "analytics.html",
}


def _route_get(handler: BaseHTTPRequestHandler, server: ViewerServer) -> None:
    """Route GET requests to API handlers or serve HTML viewer."""
    parsed = urlparse(handler.path)
    path = parsed.path
    query = parse_qs(parsed.query)

    if path in _PAGE_MAP:
        _serve_shared_file(handler, _PAGE_MAP[path], server.workspace)
    elif path.startswith("/modules/") or path.startswith("/config/"):
        _serve_subdir_static(handler, path, server.workspace)
    elif path.endswith((".css", ".js")) and not path.startswith("/api"):
        _serve_static(handler, path, server.workspace)
    elif path == "/api/health":
        handle_health(handler, server.memory_engine, server.workspace)
    elif path.startswith("/api/models"):
        handle_model_route(path, query, handler, server.model_manager, method="GET")
    elif path.startswith("/api/kb"):
        if not handle_ux_route(path, query, handler, server.memory_engine):
            handle_kb_viewer_route(path, query, handler, server.memory_engine)
    elif path.startswith("/api/memory"):
        handle_api_route(
            path, query, handler,
            server.memory_engine, server.knowledge_graph,
        )
    else:
        _send_404(handler)


def _route_post(handler: BaseHTTPRequestHandler, server: ViewerServer) -> None:
    """Route POST requests."""
    parsed = urlparse(handler.path)
    path = parsed.path

    if path == "/api/memory/ingest-file":
        handle_ingest_file_route(handler, server.memory_engine, server.workspace)
    elif path.startswith("/api/models"):
        handle_model_route(path, {}, handler, server.model_manager, method="POST")
    elif path.startswith("/api/kb"):
        handle_ux_route(path, {}, handler, server.memory_engine, method="POST")
    else:
        _send_404(handler)


def _serve_shared_file(
    handler: BaseHTTPRequestHandler, filename: str, workspace: str
) -> None:
    """Serve file from shared/viewer/. Returns error page if missing."""
    file_path = _resolve_shared_path(workspace, filename)
    if not file_path:
        _serve_viewer_error(handler, filename, workspace)
        return
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            body = f.read().encode("utf-8")
        ct = "text/html" if filename.endswith(".html") else "text/plain"
        handler.send_response(200)
        handler.send_header("Content-Type", f"{ct}; charset=utf-8")
        handler.send_header("Access-Control-Allow-Origin", "*")
        handler.send_header("Content-Length", str(len(body)))
        handler.end_headers()
        handler.wfile.write(body)
    except OSError:
        _serve_viewer_error(handler, filename, workspace)


def _serve_subdir_static(
    handler: BaseHTTPRequestHandler, path: str, workspace: str
) -> None:
    """Serve static files from shared/viewer subdirectories."""
    if ".." in path:
        _send_404(handler)
        return
    rel_path = path.lstrip("/")
    file_path = _resolve_shared_path(workspace, rel_path)
    if not file_path:
        _send_404(handler)
        return
    ext = os.path.splitext(file_path)[1]
    content_types = {".js": "application/javascript", ".css": "text/css", ".json": "application/json"}
    content_type = content_types.get(ext, "text/plain")
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            body = f.read().encode("utf-8")
        handler.send_response(200)
        handler.send_header("Content-Type", f"{content_type}; charset=utf-8")
        handler.send_header("Access-Control-Allow-Origin", "*")
        handler.send_header("Content-Length", str(len(body)))
        handler.end_headers()
        handler.wfile.write(body)
    except OSError:
        _send_404(handler)


def _serve_static(
    handler: BaseHTTPRequestHandler, path: str, workspace: str
) -> None:
    """Serve static CSS/JS files from shared/viewer directory."""
    filename = os.path.basename(path)
    if ".." in filename or "/" in filename:
        _send_404(handler)
        return
    file_path = _resolve_shared_path(workspace, filename)
    if not file_path:
        _send_404(handler)
        return
    content_type = "text/css" if filename.endswith(".css") else "application/javascript"
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            body = f.read().encode("utf-8")
        handler.send_response(200)
        handler.send_header("Content-Type", f"{content_type}; charset=utf-8")
        handler.send_header("Access-Control-Allow-Origin", "*")
        handler.send_header("Content-Length", str(len(body)))
        handler.end_headers()
        handler.wfile.write(body)
    except OSError:
        _send_404(handler)


def _resolve_shared_path(workspace: str, rel_path: str) -> str | None:
    """Resolve path within shared/viewer/. Returns None if not found."""
    if not workspace:
        return None
    full_path = os.path.join(workspace, "shared", "viewer", rel_path)
    return full_path if os.path.isfile(full_path) else None


def _serve_viewer_error(
    handler: BaseHTTPRequestHandler, filename: str, workspace: str
) -> None:
    """Error page when shared/viewer/ files are missing."""
    html = (
        "<!DOCTYPE html><html><head><title>Viewer Unavailable</title></head>"
        "<body style='background:#0f172a;color:#e2e8f0;font-family:system-ui;padding:2rem'>"
        "<h1>Viewer Unavailable</h1>"
        f"<p>shared/viewer/{filename} not found. Please ensure workspace is correct.</p>"
        f"<p style='opacity:.6;font-size:.8rem'>Workspace: {workspace or '(not set)'}</p>"
        "</body></html>"
    )
    body = html.encode("utf-8")
    handler.send_response(503)
    handler.send_header("Content-Type", "text/html; charset=utf-8")
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Content-Length", str(len(body)))
    handler.end_headers()
    handler.wfile.write(body)


def _handle_cors_preflight(handler: BaseHTTPRequestHandler) -> None:
    """Handle CORS preflight OPTIONS request."""
    handler.send_response(204)
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
    handler.send_header("Access-Control-Allow-Headers", "Content-Type")
    handler.end_headers()


def _send_404(handler: BaseHTTPRequestHandler) -> None:
    """Send 404 response."""
    body = b'{"error":"Not found"}'
    handler.send_response(404)
    handler.send_header("Content-Type", "application/json")
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Content-Length", str(len(body)))
    handler.end_headers()
    handler.wfile.write(body)


def _log(msg: str) -> None:
    print(f"[http] {msg}", file=sys.stderr, flush=True)
